infrastructure/src/main/java/org/corfudb/infrastructure/health/HealthStatus.java

```diff
@@ -0,0 +1,108 @@
+package org.corfudb.infrastructure.health;
+
+import lombok.Getter;
+import lombok.ToString;
+
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * HealthStatus of a component. It's separated into initHealthIssues and runtimeHealthIssues.
+ * initHealthIssues are present only if the component is not initialized, no runtimeHealthIssues
+ * are present at that time. Once initHealthIssues is empty (component is initialized),
+ * runtimeHealthIssues can appear.
+ */
+@ToString
+public class HealthStatus {
+
+    @Getter
+    private final Set<Issue> initHealthIssues;
+    @Getter
+    private final Set<Issue> runtimeHealthIssues;
+
+    private InitStatus initStatus = InitStatus.UNKNOWN;
+
+    public HealthStatus() {
+        this.initHealthIssues = new LinkedHashSet<>();
+        this.runtimeHealthIssues = new LinkedHashSet<>();
+    }
+
+    /**
+     * Add init health issue -- health monitor now expects the component to be initialized.
+     * If component has been not been previously initialized (UNKNOWN) or initialized (INITIALIZED),
+     * adding init issue will move it to INITIALIZED, all runtime issues are cleared.
+     * @param issue An init issue
+     */
+    public void addInitHealthIssue(Issue issue) {
+        // If component has been not been previously initialized (UNKNOWN) or
+        // initialized (INITIALIZED), adding init issue will move it to INITIALIZED,
+        // all runtime issues are cleared
+        if (initStatus == InitStatus.UNKNOWN || initStatus == InitStatus.INITIALIZED) {
+            initStatus = InitStatus.NOT_INITIALIZED;
+            runtimeHealthIssues.clear();
+        }
+        initHealthIssues.add(issue);
+    }
+
+    /**
+     * Remove the init health issue -- report to health monitor that the component is initialized.
+     * @param issue An init issue
+     */
+    public void resolveInitHealthIssue(Issue issue) {
+        if (initStatus == InitStatus.NOT_INITIALIZED) {
+            initStatus = InitStatus.INITIALIZED;
+        }
+        initHealthIssues.remove(issue);
+    }
+
+    /**
+     * Add runtime health issue -- health monitor now expects the component's runtime issue to recover.
+     * If component has been not been previously initialized (INITIALIZED), adding the runtime issue will
+     * result in exception.
+     * @param issue An init issue
+     */
+    public void addRuntimeHealthIssue(Issue issue) {
+        if (initStatus != InitStatus.INITIALIZED) {
+            throw new IllegalStateException("Runtime health issue can only be reported if the component is initialized");
```

Comment:
It is up to the caller to ensure that the runtime issues are reported after the initialization issues. A components thread can not report the runtime issue before the initialization issue because it's not started yet (Sequencer/Management Server/Layout Server do not perform any runtime activity before they are bootstrapped, for example), so if a component reports a runtime issue when it's not initialized for one reason or another, it's an illegal state and should be handled by the caller